Create rds-monitoring chart

iuriaranda opened the pull request:
This basically is a wrapper chart to set some default values for cloudwatch-exporter.
It's good to go, the only needed value is the IAM role ARN to use.

As per https://github.com/skyscrapers/engineering/issues/197

iuriaranda committed:
Create rds-monitoring chart

This basically is a wrapper chart to set some default values for cloudwatch-exporter.
It's good to go, the only needed value is the IAM role ARN to use.

As per https://github.com/skyscrapers/engineering/issues/197

simonrondelez commented on rds-monitoring/values.yaml:
```diff
@@ -0,0 +1,65 @@
+
+  config: |-
+    region: eu-west-1
+    period_seconds: 60
```
this interval might become costly -> to evaluate

duboisph replied:
It's not the exporter's period_seconds that makes it costly, but the ServiceMonitor's scraping `interval` higher up.

Still valid comment though, as that's also set to 60 😅

simonrondelez commented on rds-monitoring/values.yaml:
```diff
@@ -0,0 +1,65 @@
+      - aws_namespace: AWS/RDS
+        aws_metric_name: WriteIOPS
+        aws_dimensions: [DBInstanceIdentifier]
+        aws_statistics: [Minimum, Maximum, Average, Sum]
```
i'm missing replica lag

i'm also not sure if all of these are compatible with Aurora

iuriaranda replied:
I've added the ReplicaLag metric.
These metrics are fetched "as is" from cloudwatch, so if an RDS instance doesn't generate a specific metric, it simply won't be fetched by the exporter.

iuriaranda committed:
Complete rds-monitoring chart

- Add some prometheus alerts
- move dependencies to Chart.yaml and generate a Chart.lock
- Only fetch the necessary metrics from Cloudwatch
- Also fetch replica lag metric

iuriaranda commented:
@duboisph @simonrondelez can you re-review? I've added some alerts as well and made some changes.

iuriaranda commented:
Btw, I've deployed the chart in our cluster, in case you want to check out the alerts and metrics.

iuriaranda committed:
Remove appVersion from rds monitoring chart

We don't deploy any actual application in this chart

iuriaranda committed:
Remove useless CPUUsageHigh alert

The metric used for the alert (node_cpu) is not available anymore, so the alert doesn't do anything atm.
The fact that we didn't notice this, and that it's an info alert, tells me that it's not really useful, so I think it's better to remove it.

iuriaranda committed:
Update cloudwatch-exporter in all charts

iuriaranda commented:
Also, I'm a bit worried about how we monitor AWS resources with the cloudwatch exporter at the moment. Because like this, we're monitoring all RDS instances in the aws account, both production and staging, and this alerts will trigger regardless of the environment, and I see no way to fix that. Not sure if that's acceptable.

duboisph commented:
@iuriaranda Yeah, valid concenr and a tricky one. Afaik for monitoring SQS it's only deployed to production for some customers (no alerts I think).

But to filter RDS here, you should be able to filter on a Dimension normally. Eg we do so for ElasticSearch too: https://github.com/skyscrapers/terraform-awselasticsearch/blob/master/elasticsearch_k8s_monitoring/templates/elasticsearch-monitoring-values.yaml.tpl#L30-L31. Unfortunately there it is per cluster, not ideal either (lucky just very limited metrics we request).

Maybe it's possible to filter out via AWS tags, so we can monitor eg per environment instead?

duboisph reviewed (dismissed):
Apart from the above concerns, the PR itself looks good to me.

simonrondelez commented on rds-monitoring/templates/prometheusrules.yaml:
```diff
@@ -0,0 +1,60 @@
+            group: persistence
+          annotations:
+            message: 'CPU usage for RDS instance {{`{{ $labels.dbinstance_identifier }}`}} is higher than 95%.'
+            runbook_url: 'https://github.com/skyscrapers/documentation/tree/master/runbook.md#alert-name-rdscpuusagehigh'
```
I would add an alert for replicalag:
```
expr: aws_rds_replicalag_average{job="{{ .Release.Name }}-prometheus-cloudwatch-exporter"} > 30
```

iuriaranda replied:
done

iuriaranda committed:
Add replicalag alert

iuriaranda commented:
Ok, after some digging I've found a couple of ways we could filter environment resources:

- we can configure the cloudwatch-exporter to export an "info" metric to Prometheus, one that combines the resource id with all its AWS tags. This way we can filter production resources in the prometheus alert expression.
- we can configure the cloudwatch-exporter to only export metrics for production resources via tag filter. This way just the production resource metrics would end up in Prometheus and we won't need to filter in the alert expression. Although I think it could still be useful to have non-production resource metrics in Prometheus in any case.

Cost-wise, there shouldn't be any change as the number of API calls will be the same afaik

iuriaranda committed:
Add an additional info metric for each db

With this, cw exporter will export an additional info metric for each RDS instance with the aws tags. So we can filter for production dbs in alert expressions

iuriaranda committed:
Apply RDS alerts to only production databases

Like this it's not configurable, but afaik we only want production alerts for now.

iuriaranda commented:
I've used the tag filtering option in cloudwatch exporter to only alert on metrics from production DBs. The metrics for all instances will still end up in Prometheus in any case, but they'll be filtered out in the alert expressions.

@duboisph and @simonrondelez can you review again?

duboisph reviewed (commented):
Having the tag filter hardcoded in the alerts makes it not really dynamic. Ok if you use our terraform-rds module, but in any case tagging can be different.

Can you instead use a template for it so we can override?

Also need to fix the chart version conflict :P

iuriaranda committed:
Customize tag filtering

iuriaranda committed:
Merge remote-tracking branch 'origin/master' into rds-monitoring

iuriaranda committed:
Bump cluster-monitoring chart version

iuriaranda commented:
@duboisph I've made it configurable, and you can now also disable the filtering altogether.
